sw/device/tests/aon_timer_wakeup_irq_test.c

```diff
@@ -0,0 +1,162 @@
+#include "sw/device/lib/testing/test_framework/ottf.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+
+static const uint32_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+static const uint32_t kTickFreqHz = 1000 * 1000;  // 1Mhz / 1us
+static const uint32_t kSleepTime_us = 10 * 1000;  // 10ms
+static dif_aon_timer_t aon_timer;
+static dif_rv_timer_t rv_timer;
+static dif_rv_plic_t plic;
+
+static volatile dif_aon_timer_irq_t irq;
+static volatile top_earlgrey_plic_peripheral_t peripheral;
+static volatile uint64_t time_elapsed;
+
+/** Initialize the rv timer to count the tick
+ */
+static void tick_init(void) {
+  CHECK_DIF_OK(dif_rv_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_TIMER_BASE_ADDR), &rv_timer));
+
+  CHECK_DIF_OK(dif_rv_timer_reset(&rv_timer));
+
+  // Compute and set tick parameters (i.e., step, prescale, etc.).
+  dif_rv_timer_tick_params_t tick_params;
+  CHECK_DIF_OK(dif_rv_timer_approximate_tick_params(kClockFreqPeripheralHz,
+                                                    kTickFreqHz, &tick_params));
+
+  CHECK_DIF_OK(
+      dif_rv_timer_set_tick_params(&rv_timer, kPlicTarget, tick_params));
+
+  CHECK_DIF_OK(dif_rv_timer_counter_set_enabled(&rv_timer, kPlicTarget,
+                                                kDifToggleEnabled));
+}
+
+/** Read the current rv timer count/tick
+ * @return The current rv timer count.
+ */
+static uint64_t tick_count_get(void) {
+  uint64_t tick = 0;
+  dif_rv_timer_counter_read(&rv_timer, kPlicTarget, &tick);
+  return tick;
+}
+
+/** Execute the wake up interrupt test
+ */
+static void execute_test(dif_aon_timer_t *aon_timer, uint32_t sleep_time_us) {
+  peripheral = kTopEarlgreyPlicPeripheralUnknown;
+  irq = kDifAonTimerIrqWdogTimerBark;
+  // The wake up time should be `sleepTime_us ±15%`
+  uint32_t sleep_range_h = (float)sleep_time_us * 1.15;
+  uint32_t sleep_range_l = (float)sleep_time_us * 0.85;
+
+  // Setup the timer and wait for the IRQ
+  uint32_t sleep_cycles = (sleep_time_us * kClockFreqPeripheralHz / 1000000);
+  LOG_INFO("Setting wakeup interrupt for %u us (%u cycles)", sleep_time_us,
+           sleep_cycles);
+  // Capture the current tick to measure the time the IRQ will take
+  time_elapsed = tick_count_get();
+  aon_timer_testutils_wakeup_config(aon_timer, sleep_cycles);
+  do {
+    wait_for_interrupt();
+  } while (peripheral != kTopEarlgreyPlicPeripheralAonTimerAon &&
+           time_elapsed < sleep_range_h);
+
+  CHECK(time_elapsed < sleep_range_h && time_elapsed > sleep_range_l,
+        "Timer took %u usec which is not in the range %u usec and %u usec",
+        (uint32_t)time_elapsed, sleep_range_l, sleep_range_h);
+
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralAonTimerAon,
+        "Interrupt from incorrect peripheral: exp = %d, obs = %d",
+        kTopEarlgreyPlicPeripheralAonTimerAon, peripheral);
```

Comment:
This check will never trip due to the above time range check and while condition. In other words, it will never exit the while loop on non-aon interrupt. When the time is reached, it will, but then will get captured by the above check.

Not entirely sure what to do here, I would probably leave it as it is, or try to restructure the code.

Reply:
You are right, but I think I'll leave it this way.